src/main/java/jcifs/smb/SpnegoContext.java

```diff
@@ -0,0 +1,497 @@
+        if ( ( mechanismListMIC == null || !this.mechContext.supportsIntegrity() ) && this.requireMic
+                && !this.mechContext.isPreferredMech(this.selectedMech) ) {
+            throw new CIFSException("SPNEGO integrity is required but not available");
+        }
+
+        // otherwise we ignore the absence of a MIC
+        if ( !this.mechContext.isMICAvailable() || mechanismListMIC == null ) {
+            return;
+        }
+
+        try {
+            ASN1ObjectIdentifier[] lm = this.mechs;
+            byte[] ml = encodeMechs(lm);
+            if ( log.isInfoEnabled() ) {
+                log.debug("In Mech list " + Arrays.toString(lm));
+                log.debug("In Mech list encoded " + Hexdump.toHexString(ml));
+                log.debug("In Mech list MIC " + Hexdump.toHexString(mechanismListMIC));
+            }
+            this.mechContext.verifyMIC(ml, mechanismListMIC);
+        }
+        catch ( CIFSException e ) {
+            throw new CIFSException("Failed to verify mechanismListMIC", e);
+        }
+    }
+
+
+    /**
+     * @param mechs
+     * @return
+     * @throws CIFSException
+     */
+    private static byte[] encodeMechs ( ASN1ObjectIdentifier[] mechs ) throws CIFSException {
+        try {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream();
+            DEROutputStream dos = new DEROutputStream(bos);
+            dos.writeObject(new DERSequence(mechs));
+            dos.close();
+            return bos.toByteArray();
+        }
+        catch ( IOException e ) {
+            throw new CIFSException("Failed to encode mechList", e);
+        }
+    }
+
+
+    private SpnegoToken initialToken () throws CIFSException {
+        byte[] mechToken = this.mechContext.initSecContext(new byte[0], 0, 0);
+        return new NegTokenInit(this.mechs, this.mechContext.getFlags(), mechToken, null);
+    }
+
+
+    @Override
+    public boolean isEstablished () {
+        return this.completed && this.mechContext.isEstablished();
+    }
+
+
+    private static SpnegoToken getToken ( byte[] token, int off, int len ) throws SpnegoException {
+        byte[] b = new byte[len];
+        if ( off == 0 && token.length == len ) {
+            b = token;
+        }
+        else {
+            System.arraycopy(token, off, b, 0, len);
+        }
+        return getToken(b);
+    }
+
+
+    private static SpnegoToken getToken ( byte[] token ) throws SpnegoException {
+        SpnegoToken spnegoToken = null;
+        try {
+            switch ( token[ 0 ] ) {
+            case (byte) 0x60:
+                spnegoToken = new NegTokenInit(token);
+                break;
+            case (byte) 0xa1:
+                spnegoToken = new NegTokenTarg(token);
+                break;
+            default:
+                throw new SpnegoException("Invalid token type");
+            }
+            return spnegoToken;
+        }
+        catch ( IOException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)